src/python/WMComponent/TaskArchiver/TaskArchiverPoller.py

```diff
@@ -470,6 +463,47 @@ def notifyWorkQueue(self, subList):
 
         return
 
+    
+    def deleteWorkflowFromWMBSAndDisk(self):
+        #Get the finished workflows, in descending order
+        deletableWorkflowsDAO = self.daoFactory(classname = "Workflow.GetDeletableWorkflows")
+        deletablewfs = deletableWorkflowsDAO.execute()
+
+        #Only delete those where the upload and notification succeeded
+        logging.info("Found %d candidate workflows for deletion: %s" % (len(deletablewfs),deletablewfs.keys()))
+        # update the completed flag in dbsbuffer_workflow table so blocks can be closed
+        # create updateDBSBufferWorkflowComplete DAO
+        if len(deletablewfs) ==0:
+            return
+        safeStatesToDelete = ["completed", "aborted-completed", "rejected", 
+                              "normal-archived", "aborted-archived", "rejected-archived"]
+        wfsToDelete = {}
+        for workflow in deletablewfs:
+            try:
+                spec = retrieveWMSpec(wmWorkloadURL = deletablewfs[workflow]["spec"])
+                if not spec:
+                    raise Exception(msg = "Couldn't load spec for %s" % workflow)
+                # check workfow is alredy completed or archived (which means
+                
+                #This is used both tier0 and normal agent case
+                result = self.centralCouchDBWriter.getStatusAndTypeByRequest(workflow)
+                wfStatus = result[workflow][0]
+                if wfStatus in safeStatesToDelete:
+                    wfsToDelete[workflow] = {"spec" : spec, "workflows": deletablewfs[workflow]["workflows"]}
```

Comment:
Sorry, I didn't follow this. deletablewfs is dict with {workflow_name: {'workflows': {workflow id: [....]}
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMBS/MySQL/Workflow/GetFinishedWorkflows.py#L60

I didn't want to change the return format since it gets feed to other method.